cabby/geo/map_processing/map_structure.py

```diff
@@ -102,41 +107,194 @@ def closest_nodes(self, point: Point) -> int:
 
-    tags = {'name': True}
+    tags = {'name': True, 'building': True, 'amenity': True}
 
     osm_poi = ox.pois.pois_from_polygon(self.polygon_area, tags=tags)
+    osm_poi = osm_poi.set_crs(epsg=OSM_CRS, allow_override=True)
+
     osm_poi_named_entities = osm_poi[osm_poi['name'].notnull()]
     osm_highway = osm_poi_named_entities['highway']
     osm_poi_no_streets = osm_poi_named_entities[osm_highway.isnull()]
     osm_poi_streets = osm_poi_named_entities[osm_highway.notnull()]
 
-    osm_poi_no_streets = osm_poi_no_streets.set_crs(
-      epsg=OSM_CRS, allow_override=True)
     # Get centroid for POI.
-    osm_poi_no_streets = osm_poi_no_streets.set_crs(epsg=OSM_CRS,
-                            allow_override=True)
     osm_poi_no_streets['centroid'] = osm_poi_no_streets['geometry'].apply(
       lambda x: x if isinstance(x, Point) else x.centroid)
 
     return osm_poi_no_streets, osm_poi_streets
 
-  def get_s2cellids_for_poi(
-      self, geometry: Any) -> Optional[Sequence[s2.S2CellId]]:
-    '''get cellids for POI. 
+  def add_single_poi_to_graph(
+      self, single_poi: pd.Series) -> Sequence[edge.Edge]:
+    '''Add single POI to nx_graph.
     Arguments:
-      geometry: The geometry to which a cellids will be retrived.
+      single_poi: a single POI to be added to graph.
     Returns:
-      The the cellids.
+      The new osmid.
     '''
-    ""
+
+    # Project POI on to the closest edge in graph.
+    geometry = single_poi['geometry']
     if isinstance(geometry, Point):
-      return util.cellid_from_point(geometry, self.level)
+      points = [single_poi['geometry']]
+    elif isinstance(geometry, Polygon):
+      coords = single_poi['geometry'].exterior.coords
+      n_points = len(coords)
+
+      # Sample maximum 4 points.
+      sample_1 = Point(coords[0])
+      sample_2 = Point(coords[round(n_points/4)])
+      sample_3 = Point(coords[round(n_points/2)])
+      sample_4 = Point(coords[round(3*n_points/4)])
+      points = [sample_1, sample_2, sample_3, sample_4]
+      points = points[0:4]
     else:
-      return util.cellid_from_polygon(geometry, self.level)
+      return single_poi['osmid']
+
+    poi_osmid = single_poi['osmid']
+    poi_osmid = util.concat_numbers(9999, poi_osmid)
+    single_poi['osmid'] = poi_osmid
+    assert poi_osmid not in self.poi['osmid'].tolist(), poi_osmid
+
+    list_edges_connected_ids = []
+    eddges_to_add = []
+    for point in points:
+      eddges_to_add += self.add_single_point_edge(
+        point, list_edges_connected_ids, poi_osmid)
+
+    # Add node POI to graph.
+    self.nx_graph.add_node(
+      poi_osmid,
+      highway="poi",
+      osmid=poi_osmid,
+      x=point.x,
+      y=point.y
+    )
+
+    return eddges_to_add
+
+  def add_single_point_edge(self, point: Point,
+                list_edges_connected_ids: List, poi_osmid: int) -> Optional[Sequence[edge.Edge]]:
+    '''Connect a poi to the closest edge.
```

Comment:
It seems like this function (1) adds nodes, (2) adds edges, and (3) returns edges "to be added". I can't tell what the entities involved in these steps are -- e.g., what is the difference between a "projected" POI vs a regular POI?

Other questions that come to mind: why are we connecting a POI to an edge? Aren't POIs already in the graph (and if not, what are the nodes if not POIs)? As for the return value, can we simply add the edges in this function (as the name of the function suggests) instead of returning them?

Reply:
I added - 
    '''Connect a POI to the closest edge: (1) claculate the projected point to the nearest edge; (2) add the projected node to the graph; (3) create an edge between the point of the POI and the projcted point (add to the list to be returned); (4) create two edges from the closest edge (u-v) and the projected point: (a) u-projected point; (b) v-projected point; (5) remove closest edge u-v.
    Arguments:
      point: a point POI to be connected to the closest edge.
      list_edges_connected_ids: list of edges ids already connected to the POI. 
      If the current edge found is already connected it will avoid connecting it again.
      poi_osmid: the POI  id to be connected to the edge.
    Returns:
      The edges between the POI and the closest edge found to be added to the graph.
    '''